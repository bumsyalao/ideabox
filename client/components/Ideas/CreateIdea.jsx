import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import ReactMde, { ReactMdeCommands } from 'react-mde';
import { createIdea } from '../../actions/ideaAction';


/**
 * Create Idea component
 *
 * @class CreateIdea
 * @extends {Component}
 */
class CreateIdea extends Component {
  /**
   * Creates an instance of CreateIdea.
   * @param {any} props
   *
   * @memberOf CreateIdea
   */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      description: '',
      category: '',
      access: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleValueChange = this.handleValueChange.bind(this);
  }
  /**
   * Sets the event value to the state
   * @param {object} value
   * @return {void}
   * @memberOf CreateIdea
   */
  handleValueChange = (value) => {
    this.setState({ description: value });
  }
  /**
   * Sets the event value to the state
   * @param {object} event
   * @return {void}
   * @memberOf CreateIdea
   */
  onChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
   * Makes a call to create new Idea
   * @param {object} event
   * @return {void}
   * @memberOf CreateIdea
   */
  onSubmit(event) {
    event.preventDefault();
    const accessType = $('#switch-btn').is(':checked') ? 'public' : 'private';
    const descriptionMde = this.state.description.text;
    const newIdea = {
      title: this.state.title,
      description: descriptionMde,
      category: this.state.category,
      access: accessType
    };
    this.props.createIdea(newIdea)
      .then(() => {
        Materialize.toast('Your Idea has been created', 5000, 'green');
      }).catch((err) => {
        Materialize.toast(err.response.data.message, 5000, 'red');
      });
  }
  /**
   * renders create idea component
   * @returns {void}
   *
   * @memberOf CreateIdea
   */
  render() {
    $(document).ready(function() {
      $('select').material_select();
      $('#switch-btn').prop('checked');
      $('.tooltipped').tooltip({ delay: 50 });
    });
    $('.tooltipped').tooltip('remove');
    return (
      <div className="create-idea">
        <div className="row">
          <Link
            to="/dashboard/idea-list"
            className="btn tooltipped btn-floating btn-large waves-effect waves-light red close-btn right"
            data-position="bottom"
            data-delay="50"
            data-tooltip="close"
          >
            <i
              className="material-icons"
            >close</i></Link>
          <h4> Write down that Idea! </h4>
          <form className="col s12 form-margin">
            <div className="row">
              <div className="input-field col s12 inline">
                <input
                  id="title"
                  value={this.state.title}
                  onChange={this.onChange}
                  type="text"
                  className="validate"
                />
                <label htmlFor="Title">Title</label>
              </div>
            </div>
            <div className="input-field col s12">
              <select
                className="browser-default"
                name="category"
                id="category"
                value={this.state.category}
                onChange={this.onChange}
              >
                <option value="" disabled selected>
                    Choose a category
                </option>
                <option value="Arts and Entertainment">Arts and Entertainment</option>
                <option value="Cars and other vehicles">Cars and other vehicles</option>
                <option value="Computers and Electronics">Computers and Electronics</option>
                <option value="Education and Communication">Education and Communication</option>
                <option value="Family Life">Family Life</option>
                <option value="Finance and Business">Finance and Business</option>
                <option value="Food and Entertaining">Food and Entertaining</option>
                <option value="Health">Health</option>
                <option value="Hobbies and Craft">Hobbies and Craft</option>
                <option value="Holidays and Traditions">Holidays and Traditions</option>
                <option value="Home and Gardens">Home and Gardens</option>
                <option value="Personal care and Style">Personal care and style</option>
                <option value="Pets and Aminals">Pets and Aminals</option>
                <option value="Philosophy and Religion">Philosophy and Religion</option>
                <option value="Relationships">Relationships</option>
                <option value="Sports and Finance">Sports and Finance</option>
                <option value="Travel">Travel</option>
                <option value="Work world">Work world</option>
                <option value="Youth">Youth</option>

              </select>
            </div>
            <div className="col s12">
              <ReactMde
                textAreaProps={{
                  id: 'description',
                  name: 'description',
                }}
                id="description"
                value={this.state.description}
                onChange={this.handleValueChange}
                commands={ReactMdeCommands.getDefaultCommands()}
              />
            </div>
            <div className="switch">
              <label>
                  Private Idea
                <input id="switch-btn" type="checkbox" value={this.state.access} />
                <span className="lever" />
                  Public Idea
              </label>
              <p> Public Ideas can be seen by everyone </p>
            </div>
            <button
              className="waves-effect waves-light btn right"
              onClick={this.onSubmit}
            >Create Idea</button>
          </form>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  user: state.access.user,
});

export default connect(mapStateToProps, { createIdea })(withRouter(CreateIdea));
