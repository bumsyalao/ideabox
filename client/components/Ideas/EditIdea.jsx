import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import ReactMde, { ReactMdeCommands } from 'react-mde';
import { editIdea, getUserIdeas, getIdea } from '../../actions/ideaAction';

/**
 * Edit Idea Class
 *
 * @class EditIdea
 * @extends {Component}
 */
class EditIdea extends Component {
  /**
   * Creates an instance of EditIdea.
   * @param {any} props
   *
   * @memberOf EditIdea
   */
  constructor(props) {
    super(props);
    this.state = {
      title: this.props.idea.title,
      description: this.props.idea.description || {},
      category: this.props.idea.category,
      access: this.props.idea.access
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleValueChange = this.handleValueChange.bind(this);
  }

/**
 * Makes a call to get Idea
 *
 * @return {void}
 * @memberOf EditIdea
 */
  componentWillMount() {
    const ideaId = this.props.match.params.ideaId;
    this.props.getIdea(ideaId)
      .then(() => {
        this.setState({
          title: this.props.idea.title,
          category: this.props.idea.category,
          description: this.props.idea.description,
          access: this.props.idea.access
        });
      });
  }
/**
 * Updates state if props are changed
 * @param {any} nextProps
 * @return {void}
 * @memberOf EditIdea
 */
  componentWillReceiveProps(nextProps) {
    this.setState({
      title: nextProps.idea.title,
      description: nextProps.idea.description,
      category: nextProps.idea.category,
      access: nextProps.idea.access
    });
  }
  /**
   * Sets the event value to the state
   * @param {object} event
   * @return {void}
   * @memberOf EditIdea
   */
  onChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
   * Sets the event value to the state
   * @param {object} value
   * @return {void}
   * @memberOf CreateIdea
   */
  handleValueChange = (value) => {
    this.setState({ description: value });
  }

  /**
   * Makes a call to edit Idea
   * @param {object} event
   * @return {void}
   * @memberOf EditIdea
   */
  onSubmit(event) {
    event.preventDefault();
    const { ideaId } = this.props.match.params;
    const accessType = $('#switch-btn').is(':checked') ? 'public' : 'private';
    const descriptionMde = this.state.description.text;
    const newIdea = {
      title: this.state.title,
      description: descriptionMde,
      category: this.state.category,
      access: accessType
    };
    this.props.editIdea(newIdea, ideaId)
      .then(() => {
        Materialize.toast('Your Idea has been updated', 5000, 'green');
        this.props.history.push('/dashboard/my-ideas');
      }).catch((err) => {
        Materialize.toast(err.response.data.message, 5000, 'red');
      });
  }
  /**
   * Renders Edit Idea component
   * @returns {void}
   *
   * @memberOf EditIdea
   */
  render() {
    $(document).ready(() => {
      $('select').material_select();
      $('#switch-btn').prop('checked');
      $('.tooltipped').tooltip({ delay: 50 });
    });
    $('.tooltipped').tooltip('remove');
    return (
      <div className="create-idea">
        <div className="row">
          <h4> Edit Idea </h4>
          <form className="col s12 form-margin">
            <div className="row">
              <div className="input-field col s12">
                <input
                  id="title"
                  value={this.state.title}
                  onChange={this.onChange}
                  type="text"
                  className="validate"
                />
              </div>
            </div>
            <div className="input-field col s12">
              <select
                className="browser-default"
                name="category"
                id="category"
                value={this.state.category}
                onChange={this.onChange}
              >
                <option value="" disabled selected>
                    Choose a category
                </option>
                <option value="Arts and Entertainment">Arts and Entertainment</option>
                <option value="Cars and other vehicles">Cars and other vehicles</option>
                <option value="Computers and Electronics">Computers and Electronics</option>
                <option value="Education and Communication">Education and Communication</option>
                <option value="Family Life">Family Life</option>
                <option value="Finance and Business">Finance and Business</option>
                <option value="Food and Entertaining">Food and Entertaining</option>
                <option value="Health">Health</option>
                <option value="Hobbies and Craft">Hobbies and Craft</option>
                <option value="Holidays and Traditions">Holidays and Traditions</option>
                <option value="Home and Gardens">Home and Gardens</option>
                <option value="Personal care and Style">Personal care and style</option>
                <option value="Pets and Aminals">Pets and Aminals</option>
                <option value="Philosophy and Religion">Philosophy and Religion</option>
                <option value="Relationships">Relationships</option>
                <option value="Sports and Finance">Sports and Finance</option>
                <option value="Travel">Travel</option>
                <option value="Work world">Work world</option>
                <option value="Youth">Youth</option>

              </select>
            </div>
            <div className="input-field col s12">
              <ReactMde
                textAreaProps={{
                  id: 'description',
                  name: 'description',
                }}
                id="description"
                value={this.state.description}
                onChange={this.handleValueChange}
                commands={ReactMdeCommands.getDefaultCommands()}
              />
            </div>
            <div className="switch">
              <label>
                  Private Idea
                <input id="switch-btn" type="checkbox" value={this.state.access} />
                <span className="lever" />
                  Public Idea
              </label>
              <p> Public Ideas can be seen by everyone </p>
            </div>
            <button
              className="waves-effect waves-light btn right"
              onClick={this.onSubmit}
            >Save Idea</button>
          </form>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  user: state.access.user,
  ideas: state.ideas.ideas,
  idea: state.ideas.idea,

});

export default connect(mapStateToProps, { editIdea, getUserIdeas, getIdea })(
  withRouter(EditIdea));
