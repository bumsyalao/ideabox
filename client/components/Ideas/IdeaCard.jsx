import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import * as Showdown from 'showdown';

/**
 * Idea card class
 *
 * @class IdeaCard
 * @extends {Component}
 */
export class IdeaCard extends Component {

  /**
   * Creates an instance of IdeaCard.
   * @param {object} props
   *
   * @memberOf IdeaCard
   */
  constructor(props) {
    super(props);
    this.copyLink = this.copyLink.bind(this);
  }

  /**
   * Functo to copy to clipboard
   *
   * @param {string} text
   * @returns {void}
   * @memberOf IdeaCard
   */
  copyLink(text) {
    if (window.clipboardData && window.clipboardData.setData) {
      // IE specific code path to prevent textarea being shown while dialog is visible.
      return clipboardData.setData('Text', text);
    } else if (
      document.queryCommandSupported &&
      document.queryCommandSupported('copy')
    ) {
      const textarea = document.createElement('textarea');
      textarea.textContent = text;
      textarea.style.position = 'fixed'; // Prevent scrolling to bottom of page in MS Edge.
      document.body.appendChild(textarea);
      textarea.select();
      try {
        return document.execCommand('copy'); // Security exception may be thrown by some browsers.
      } catch (ex) {
        console.warn('Copy to clipboard failed.', ex);
        return false;
      } finally {
        document.body.removeChild(textarea);
        Materialize.toast('Copied to clipboard', 5000, 'green');
      }
    }
  }

  /**
   * Renders Idea card class
   * @returns {void}
   *
   * @memberOf IdeaCard
   */
  render() {
    $(document).ready(() => {
      $('.tooltipped').tooltip({ delay: 50 });
    });
    $('.tooltipped').tooltip('remove');
    const { title, description, category, authorName, id } = this.props;
    const converter = new Showdown.Converter();
    const mdDescription = converter.makeHtml(description);
    return (
      <div className="grid-item card hoverable">
        <div className="card-content">
          <span className="card-title activator grey-text text-darken-4">
            {title}
            <i className="material-icons right">more_vert</i>
          </span>
          <p> By: {authorName} </p>
          <div className="category-text">{category}</div>
        </div>
        <div className="card-action">
          <ul className="card-icons">
            <li>
              <a
                id={id}
                href={`${window.location.origin}/view/${id}`}
                onClick={(event) => {
                  event.preventDefault();
                  this.copyLink(`${window.location.origin}/view/${id}`);
                }}
              >
                <i
                  className="tiny material-icons icon-btn tooltipped"
                  data-position="bottom"
                  data-delay="50"
                  data-tooltip="copy link"
                >
                  insert_link
                </i>
              </a>
            </li>
            <li>
              <Link to={`/dashboard/view/${id}`}>
                <i
                  className="tiny material-icons icon-btn tooltipped"
                  data-position="bottom"
                  data-delay="50"
                  data-tooltip="view article"
                >
                  remove_red_eye
                </i>
              </Link>
            </li>
          </ul>
        </div>
        <div className="card-reveal">
          <span className="card-title grey-text text-darken-4">
            <i className="material-icons right">close</i>
          </span>
          <p
            dangerouslySetInnerHTML={{ __html: mdDescription }}
            className="truncate"
          />
        </div>
      </div>
    );
  }
}

export default IdeaCard;
