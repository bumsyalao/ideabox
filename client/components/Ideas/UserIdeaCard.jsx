import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import * as Showdown from 'showdown';

import { deleteIdea } from '../../actions/ideaAction';

/**
 * User Idea card component
 * @class UserIdeaCard
 * @extends {Component}
 */
export class UserIdeaCard extends Component {
  /**
   * Creates an instance of UserIdeaCard.
   * @param {any} props
   * @return {void}
   * @memberOf UserIdeaCard
   */
  constructor(props) {
    super(props);
    this.onDelete = this.onDelete.bind(this);
  }

  /**
   * Makes an action call to delete idea
   *
   * @return {void}
   * @memberOf UserIdeaCard
   */
  onDelete() {
    const ideaId = this.props.id;
    this.props.deleteIdea(ideaId);
    $('.tooltipped').tooltip('remove');
  }
  /**
   * Renders delete idea component
   *
   * @returns {void}
   *
   * @memberOf UserIdeaCard
   */
  render() {
    $(document).ready(() => {
      $('.tooltipped').tooltip({ delay: 50 });
    });
    const {
      title,
      description,
      category,
      authorName,
      modified,
      access,
      id
    } = this.props;
    const converter = new Showdown.Converter();
    const mdDescription = converter.makeHtml(description);
    return (
      <div className="grid-item card hoverable">
        <div className="card-content">
          <span className="card-title activator grey-text text-darken-4">
            {title}
            <i className="material-icons right">more_vert</i>
          </span>
          <p> By: {authorName} </p>
          <p className="right">{modified && 'Edited'}</p>
          <div className="category-text">{category}</div>
          <p>access type: {access}</p>
        </div>
        <div className="card-action">
          <ul className="card-icons">
            <li>
              <Link to={`/dashboard/edit/${id}`}>
                <i
                  className="tiny material-icons icon-btn tooltipped"
                  data-position="bottom"
                  data-delay="50"
                  data-tooltip="edit"
                >
                  edit
                </i>
              </Link>
            </li>
            <li>
              <a onClick={this.onDelete}>
                <i
                  className="tiny material-icons icon-btn tooltipped delete-btn"
                  data-position="bottom"
                  data-delay="50"
                  data-tooltip="delete"
                >
                  delete
                </i>
              </a>
            </li>
          </ul>
        </div>
        <div className="card-reveal">
          <span className="card-title grey-text text-darken-4">
            <i className="material-icons right">close</i>
          </span>
          <p
            dangerouslySetInnerHTML={{ __html: mdDescription }}
            className="truncate"
          />
        </div>
      </div>
    );
  }
}

export default connect(null, { deleteIdea })(UserIdeaCard);
