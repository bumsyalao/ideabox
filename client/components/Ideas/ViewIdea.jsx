import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as Showdown from 'showdown';
import { getIdea, addComment } from '../../actions/ideaAction';
import twitterIcon from '../../images/twitter-icon.png';

/**
 * View Idea Component
 *
 * @class ViewIdea
 * @extends {Component}
 */
class ViewIdea extends Component {
  /**
   * Creates an instance of ViewIdea.
   * @param {any} props
   *
   * @memberOf ViewIdea
   */
  constructor(props) {
    super(props);
    this.state = {
      comment: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * Makes an action to get Idea
   * @return {void}
   * @memberOf ViewIdea
   */
  componentWillMount() {
    $('.tooltipped').tooltip('remove');
    const ideaId = this.props.match.params.ideaId;
    this.props.getIdea(ideaId);
  }

  /**
   * Sets the event value to the state
   * @param {object} event
   * @return {void}
   * @memberOf CreateIdea
   */
  onChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
   * Makes an action call to add comment
   *
   * @param {any} event
   * @return {void}
   * @memberOf ViewIdea
   */
  onSubmit(event) {
    event.preventDefault();
    const ideaId = this.props.match.params.ideaId;
    const newComment = this.state.comment;
    this.props.addComment(ideaId, newComment)
    .then(() => {
      this.setState({
        comment: ''
      });
    });
  }

  /**
   * Renders View Idea component
   *
   * @return {void}
   * @memberOf ViewIdea
   */
  render() {
    $(document).ready(function() {
      $('.tooltipped').tooltip({ delay: 50 });
    });
    const converter = new Showdown.Converter();
    const mdDescription = converter.makeHtml(this.props.idea.description);
    const id = this.props.idea._id;
    return (
      <div className="dashboard">
      <div className="view-page-wrapper">
      <div className="view-page">
        <div className="row">
          <h4>{this.props.idea.title}</h4>
          <a
            href={`https://twitter.com/intent/tweet?text=This%20is%20amazing%20you%20should%20read%20it&url=${window.location.origin}/view/${id}`}
            className="tooltipped right"
            data-position="bottom"
            data-delay="50"
            data-tooltip="Like this? share on twitter"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              src={twitterIcon}
              height="50px"
              width="50px"
              className="right twitter-icon"
            />
          </a>
          <p> By: {this.props.idea.authorName} </p>
          <div className="description-body">
            <p dangerouslySetInnerHTML={{ __html: mdDescription }} />
          </div>
          <div className="comment-section">
            <ul>
              {this.props.idea.comments &&
                this.props.idea.comments.map(comment => (
                  <li key={comment._id}>
                    <p>@{comment.authorName}</p>
                    {comment.comment}
                  </li>
                ))}
            </ul>
          </div>
          <label htmlFor="textarea1">Add a comment</label>
          <textarea
            id="comment"
            value={this.state.comment}
            onChange={this.onChange}
            className="materialize-textarea"
          />
          <a onClick={this.onSubmit} className="waves-effect waves-light right">
            <i className="material-icons card-icons">send</i>
          </a>
        </div>
      </div>
      </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  idea: state.ideas.idea
});

export default connect(mapStateToProps, { getIdea, addComment })(
  ViewIdea
);
