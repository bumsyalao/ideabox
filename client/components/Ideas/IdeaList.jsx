import React, { Component } from 'react';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';
import IdeaCard from './IdeaCard';
import { getPublicIdeas, searchIdeas } from '../../actions/ideaAction';

/**
 * Idea list component
 * @class IdeaList
 * @extends {Component}
 */
export class IdeaList extends Component {
  /**
   * Creates an instance of IdeaList.
   * @param {any} props
   * @return {void}
   * @memberOf IdeaList
   */
  constructor(props) {
    super(props);
    this.state = {
      ideas: [],
      searchParam: '',
      category: ''
    };
    this.onSearch = this.onSearch.bind(this);
    this.handlePageClick = this.handlePageClick.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onCancel = this.onCancel.bind(this);
  }
  /**
   * Makes an action call to get Ideas
   * @return {void}
   * @memberOf IdeaList
   */
  componentWillMount() {
    this.props.getPublicIdeas().catch();
  }

  /**
   * Update the state if the props are changed
   * @param {object} nextProps
   * @return {void}
   * @memberOf IdeaList
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      pageCount: nextProps.pagination.pageCount,
      count: nextProps.pagination.count,
      ideas: nextProps.ideas,
      foundIdeas: nextProps.foundIdeas
    });
  }
  /**
   * Makes an action call to search for ideas
   *
   * @param {int} offset
   * @param {int} limit
   * @return {void}
   * @memberOf IdeaList
   */
  onSearch() {
    const limit = 9;
    const offset = 0;
    this.props.searchIdeas(
      offset,
      limit,
      this.state.searchParam,
      this.state.category
    )
    .then(() => this.setState({
      title: this.props.foundIdeas.title,
      id: this.props.foundIdeas._id,
      description: this.props.foundIdeas.description,
      access: this.props.foundIdeas.access,
      authorName: this.props.foundIdeas.authorName
    }));
  }
  /**
   * Sets the event value to the state
   * @param {object} event
   * @return {void}
   * @memberOf IdeaList
   */
  onChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
   * Updates state when function is called
   * @return {void}
   * @memberOf IdeaList
   */
  onCancel() {
    this.setState({
      searchParam: '',
      foundIdeas: []
    });
  }

  /**
   * Pagination for list of ideas
   * @param {object} pageData
   * @return {void}
   * @memberOf Ideas
   */
  handlePageClick(pageData) {
    const selected = pageData.selected;
    const limit = 20;
    const offset = Math.ceil(selected * limit);
    this.setState({ offset });
    this.getIdeas(offset, limit);
  }

  /**
   * Renders Idea list component
   * @returns {void}
   *
   * @memberOf IdeaList
   */
  render() {
    let result = this.state.ideas || [];

    if (this.state.foundIdeas && this.state.foundIdeas.length > 0) {
      result = this.state.foundIdeas;
    }

    return (
      <div>
        <div className="row search-row">
          <form className="col s12">
            <div className="row">
              <div className="input-field col s3 inline right white-text">
                <a
                  className="row waves-effect waves-light right cancel-btn"
                  onClick={this.onCancel}
                ><i className="material-icons icon-btn">cancel</i></a>
                <input
                  id="searchParam"
                  value={this.state.searchParam}
                  onChange={this.onChange}
                  type="text"
                  className="validate"
                />
                <label htmlFor="search-param">Search Idea</label>
              </div>
              </div>
              <div className="row input-field col s3 right">
                <select
                  className="browser-default"
                  name="category"
                  id="category"
                  value={this.state.category}
                  onChange={this.onChange}
                >
                  <option value="" disabled selected>Filter by category</option>
                  <option value="Arts and Entertainment">arts and entertainment</option>
                  <option value="Cars and other vehicles">cars and other vehicles</option>
                  <option value="Computers and Electronics">computers and electronics</option>
                  <option value="Education and Communication">education and communication</option>
                  <option value="Family Life">family life</option>
                  <option value="Finance and Business">finance and business</option>
                  <option value="Food and Entertaining">food and entertaining</option>
                  <option value="Health">health</option>
                  <option value="Hobbies and Craft">hobbies and craft</option>
                  <option value="Holidays and Traditions">holidays and traditions</option>
                  <option value="Home and Gardens">home and gardens</option>
                  <option value="Personal care and Style">personal care and style</option>
                  <option value="Pets and Aminals">pets and aminals</option>
                  <option value="Philosophy and Religion">philosophy and religion</option>
                  <option value="Relationships">relationships</option>
                  <option value="Sports and Finance">sports and finance</option>
                  <option value="Travel">travel</option>
                  <option value="Work world">work world</option>
                  <option value="Youth">youth</option>
                </select>
              <a
                className="row waves-effect waves-light right"
                onClick={this.onSearch}
              ><i className="material-icons">search</i></a>
              </div>
          </form>
        </div>
        <div className="idea-list grid grid-margin">
          {result.map(idea => (
            <IdeaCard
              key={idea._id}
              id={idea._id}
              title={idea.title}
              description={idea.description}
              category={idea.category}
              authorName={idea.authorName}
              access={idea.access}
            />
        ))}
          {(this.state.count > 9) && <ReactPaginate
            previousLabel={'previous'}
            nextLabel={'next'}
            pageCount={this.state.pageCount}
            marginPagesDisplayed={1}
            pageRangeDisplayed={3}
            onPageChange={this.handlePageClick}
            containerClassName={'pagination'}
            subContainerClassName={'pages pagination'}
            activeClassName={'active'}
          />}
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  ideas: state.ideas.ideas,
  foundIdeas: state.ideas.foundIdeas,
  pagination: state.ideas.pagination
});
export default connect(mapStateToProps, { getPublicIdeas, searchIdeas })(IdeaList);
