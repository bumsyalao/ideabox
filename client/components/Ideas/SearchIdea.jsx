import React, { Component } from 'react';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';
import { searchIdeas } from '../../actions/ideaAction';
import IdeaCard from './IdeaCard';

/**
 *
 *
 * @class SearchIdea
 * @extends {Component}
 */
class SearchIdea extends Component {

  /**
   * Creates an instance of SearchIdea.
   * @param {any} props
   *
   * @memberOf SearchIdea
   */
  constructor(props) {
    super(props);
    this.state = {
      ideas: [],
      searchParam: '',
      category: ''
    };
    this.onSearch = this.onSearch.bind(this);
    this.handlePageClick = this.handlePageClick.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  /**
   * Update the state if the props are changed
   * @param {object} nextProps
   * @return {void}
   * @memberOf UsersPage
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      ideas: nextProps.foundIdeas,
      pageCount: nextProps.pagination.pageCount,
      count: nextProps.pagination.count
    });
  }
  /**
   *
   *
   * @param {int} offset
   * @param {int} limit
   * @return {void}
   * @memberOf SearchIdea
   */
  onSearch() {
    const limit = 9;
    const offset = 0;
    this.props.searchIdeas(
      offset,
      limit,
      this.state.searchParam,
      this.state.category
    );
  }

  /**
   *
   *
   * @param {object} event
   * @return {void}
   * @memberOf CreateIdea
   */
  onChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
   * Pagination for list of users
   * @param {object} pageData
   * @return {void}
   * @memberOf UsersPage
   */
  handlePageClick(pageData) {
    const selected = pageData.selected;
    const limit = 20;
    const offset = Math.ceil(selected * limit);
    this.setState({ offset });
    this.getIdeas(offset, limit);
  }

  /**
   *
   *
   * @returns {void}
   *
   * @memberOf SearchIdea
   */
  render() {
    return (
      <div className="container">
      <div className="row">
        <form className="col s12">
          <div className="row">
            <div className="input-field col s6 inline">
              <input
                id="searchParam"
                value={this.state.searchParam}
                onChange={this.onChange}
                type="text"
                className="validate"
              />
              <label htmlFor="sear-param">Search Idea</label>
            </div>
            <div className="input-field col s6">
              <select
                className="browser-default"
                name="category"
                id="category"
                value={this.state.category}
                onChange={this.onChange}
              >
                <option value="" disabled selected>Choose a category</option>
                <option value="Arts and Entertainment">arts and entertainment</option>
                <option value="Cars and other vehicles">cars and other vehicles</option>
                <option value="Computers and Electronics">computers and electronics</option>
                <option value="Education and Communication">education and communication</option>
                <option value="Family Life">family life</option>
                <option value="Finance and Business">finance and business</option>
                <option value="Food and Entertaining">food and entertaining</option>
                <option value="Health">health</option>
                <option value="Hobbies and Craft">hobbies and craft</option>
                <option value="Holidays and Traditions">holidays and traditions</option>
                <option value="Home and Gardens">home and gardens</option>
                <option value="Personal care and Style">personal care and style</option>
                <option value="Pets and Aminals">pets and aminals</option>
                <option value="Philosophy and Religion">philosophy and religion</option>
                <option value="Relationships">relationships</option>
                <option value="Sports and Finance">sports and finance</option>
                <option value="Travel">travel</option>
                <option value="Work world">work world</option>
                <option value="Youth">youth</option>
              </select>
            </div>
            <a
              className="waves-effect waves-light btn right"
              onClick={this.onSearch}
            >Search</a>
          </div>
        </form>
      </div>
      <div className="idea-list grid">
      {this.state.ideas.map(idea => (
        <IdeaCard
          key={idea._id}
          id={idea._id}
          title={idea.title}
          description={idea.description}
          category={idea.category}
          authorName={idea.authorName}
          access={idea.access}
        />
      ))}
        {(this.state.count > 9) && <ReactPaginate
          previousLabel={'previous'}
          nextLabel={'next'}
          pageCount={this.state.pageCount}
          marginPagesDisplayed={1}
          pageRangeDisplayed={3}
          onPageChange={this.handlePageClick}
          containerClassName={'pagination'}
          subContainerClassName={'pages pagination'}
          activeClassName={'active'}
        />}
      </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  foundIdeas: state.ideas.foundIdeas,
  pagination: state.ideas.pagination
});


export default connect(mapStateToProps, { searchIdeas })(SearchIdea);
