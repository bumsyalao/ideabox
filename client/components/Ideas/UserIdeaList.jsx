import React, { Component } from 'react';
import { connect } from 'react-redux';
import UserIdeaCard from './UserIdeaCard';
import { getUserIdeas } from '../../actions/ideaAction';

/**
 * User Idea Component
 * @class UserIdeaList
 * @extends {Component}
 */
class IdeaList extends Component {

  /**
   * Makes an action call to get user ideas
   * @return {void}
   * @memberOf UserIdeaList
   */
  componentWillMount() {
    this.props.getUserIdeas().catch();
  }

  /**
   * Renders User Idea list
   * @returns {void}
   *
   * @memberOf UserIdeaList
   */
  render() {
    return (
      <div className="idea-list grid">
        {this.props.ideas.map(idea => (
          <UserIdeaCard
            key={idea._id}
            id={idea._id}
            title={idea.title}
            description={idea.description}
            category={idea.category}
            authorName={idea.authorName}
            access={idea.access}
            modified={idea.modified}
          />
        ))}
      </div>
    );
  }
}
const mapStateToProps = state => ({
  ideas: state.ideas.myIdeas
});
export default connect(mapStateToProps, { getUserIdeas })(IdeaList);
