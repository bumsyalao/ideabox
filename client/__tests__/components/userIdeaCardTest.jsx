/* global expect jest test */
import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import adapter from 'enzyme-adapter-react-16';
import { UserIdeaCard } from '../../components/Ideas/UserIdeaCard';

configure({ adapter: new adapter() });

describe('List Idea component', () => {
  const spy = jest.fn();
  beforeEach(() => {
    global.Materialize = { toast: jest.fn() };
  });
  afterEach(() => {
    spy.mockReset();
  });
  const props = {
    deleteIdea: jest.fn(() => Promise.resolve())
  };
  const component = shallow(
    <UserIdeaCard {...props} />
  );
  it('should match the Idea List Snapshot', () => {
    const tree = toJson(component);
    expect(component.find('IdeaList').length).toEqual(0);
    expect(tree).toMatchSnapshot();
  });

  it('should render component and subcomponent', () => {
    expect(component.exists()).toBe(true);
    expect(component.exists()).toBe(true);
  });

  it('calls onDelete function', () => {
    const onDeleteSpy = jest.spyOn(
        component.instance(), 'onDelete'
      );
    const ideaId = '5a312d5c01ce62169e8900d3';
    component.instance().onDelete(ideaId);
    expect(onDeleteSpy).toHaveBeenCalled();
  });
});
