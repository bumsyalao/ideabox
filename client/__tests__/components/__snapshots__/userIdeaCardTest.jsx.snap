// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Idea component should match the Idea List Snapshot 1`] = `
<div
  className="grid-item card hoverable"
>
  <div
    className="card-content"
  >
    <span
      className="card-title activator grey-text text-darken-4"
    >
      <i
        className="material-icons right"
      >
        more_vert
      </i>
    </span>
    <p>
       By: 
       
    </p>
    <p
      className="right"
    />
    <div
      className="category-text"
    />
    <p>
      access type: 
    </p>
  </div>
  <div
    className="card-action"
  >
    <ul
      className="card-icons"
    >
      <li>
        <Link
          replace={false}
          to="/dashboard/edit/undefined"
        >
          <i
            className="tiny material-icons icon-btn tooltipped"
            data-delay="50"
            data-position="bottom"
            data-tooltip="edit"
          >
            edit
          </i>
        </Link>
      </li>
      <li>
        <a
          onClick={[Function]}
        >
          <i
            className="tiny material-icons icon-btn tooltipped delete-btn"
            data-delay="50"
            data-position="bottom"
            data-tooltip="delete"
          >
            delete
          </i>
        </a>
      </li>
    </ul>
  </div>
  <div
    className="card-reveal"
  >
    <span
      className="card-title grey-text text-darken-4"
    >
      <i
        className="material-icons right"
      >
        close
      </i>
    </span>
    <p
      className="truncate"
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
  </div>
</div>
`;
